Mathlib/Data/Nat/NormNum.lean

```diff
@@ -0,0 +1,251 @@
+def FactorsHelper (n p : ℕ) (l : List ℕ) : Prop :=
+  p.Prime → List.Chain (· ≤ ·) p l ∧ (∀ a ∈ l, Nat.Prime a) ∧ List.prod l = n
+
+theorem factorsHelper_nil (a : ℕ) : FactorsHelper 1 a [] := fun _ =>
+  ⟨List.Chain.nil, by rintro _ ⟨⟩, List.prod_nil⟩
+
+theorem factorsHelper_cons' (n m a b : ℕ) (l : List ℕ) (h₁ : b * m = n) (h₂ : a ≤ b)
+    (h₃ : minFac b = b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) := fun pa =>
+  have pb : b.Prime := Nat.prime_def_minFac.2 ⟨le_trans pa.two_le h₂, h₃⟩
+  let ⟨f₁, f₂, f₃⟩ := H pb
+  ⟨List.Chain.cons h₂ f₁,
+    fun c h => (List.eq_or_mem_of_mem_cons h).elim (fun e => e.symm ▸ pb) (f₂ _),
+    by rw [List.prod_cons, f₃, h₁]⟩
+
+theorem factorsHelper_cons (n m a b : ℕ) (l : List ℕ) (h₁ : b * m = n) (h₂ : a < b)
+    (h₃ : minFac b = b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) :=
+  factorsHelper_cons' _ _ _ _ _ h₁ h₂.le h₃ H
+
+theorem factorsHelper_sn (n a : ℕ) (h₁ : a < n) (h₂ : minFac n = n) : FactorsHelper n a [n] :=
+  factorsHelper_cons _ _ _ _ _ (mul_one _) h₁ h₂ (factorsHelper_nil _)
+
+theorem factorsHelper_same (n m a : ℕ) (l : List ℕ) (h : a * m = n) (H : FactorsHelper m a l) :
+    FactorsHelper n a (a :: l) := fun pa =>
+  factorsHelper_cons' _ _ _ _ _ h le_rfl (Nat.prime_def_minFac.1 pa).2 H pa
+
+theorem factorsHelper_same_sn (a : ℕ) : FactorsHelper a a [a] :=
+  factorsHelper_same _ _ _ _ (mul_one _) (factorsHelper_nil _)
+
+theorem factorsHelper_end (n : ℕ) (l : List ℕ) (H : FactorsHelper n 2 l) : Nat.factors n = l :=
+  let ⟨h₁, h₂, h₃⟩ := H Nat.prime_two
+  have := List.chain'_iff_pairwise.1 (@List.Chain'.tail _ _ (_ :: _) h₁)
+  (List.eq_of_perm_of_sorted (Nat.factors_unique h₃ h₂) this (Nat.factors_sorted _)).symm
+
+open NormNum
+
+/-- Produce a proof that `n` is not prime from a factor `1 < d < n`. -/
+def deriveNotPrime (n d : ℕ) : MetaM <| Q(¬ Nat.Prime $n) := do
+  let d' : ℕ := n / d
+  let prf : Q($d * $d' = $n) := (q(Eq.refl $n) : Expr)
+  let r : Q(Nat.ble $d 1 = false) := (q(Eq.refl false) : Expr)
+  let r' : Q(Nat.ble $d' 1 = false) := (q(Eq.refl false) : Expr)
+  return q(Nat.not_prime_mul' $prf (isNat_lt_true (.raw_refl _) (.raw_refl _) $r)
+    (isNat_lt_true (.raw_refl _) (.raw_refl _) $r'))
+
+/-- The `norm_num` extension which identifies expressions of the form `minFac n`. -/
+@[norm_num Nat.minFac _] partial def evalMinFac :
+  NormNumExt where eval {u α} e := do
+  let .app (.const `Nat.minFac _) (n : Q(ℕ)) ← whnfR e | failure
+  let sℕ : Q(AddMonoidWithOne ℕ) := q(instAddMonoidWithOneNat)
+  let ⟨nn, pn⟩ ← deriveNat n sℕ
+  let n' := nn.natLit!
+  let rec aux (k : ℕ) (prf : Q(MinFacHelper $nn $k)) :
+    MetaM <| (c : Q(ℕ)) × Q(IsNat (Nat.minFac $n) $c) := do
+    -- remark: `deriveBool q($nn < $k * $k)` is 2x slower than the following test.
+    if n' < k * k then
+      let r : Q(Nat.ble ($k * $k) $nn = false) := (q(Eq.refl false) : Expr)
+      return ⟨nn, q(isNat_minFac_4 $pn $prf (isNat_lt_true (.raw_refl _) (.raw_refl _) $r))⟩
```

Comment:
these lemmas don't look properly calibrated; if done correctly you should never have to put together more than one lemma application per `q(...)` (except possibly for `Eq.refl` subproofs). Certainly there is no need to be passing instances, since `minFac` is not generic